src/sentry/search/django/backend.py

```diff
@@ -8,275 +8,397 @@
-                tags,
-            )
-            if not matches:
-                return queryset.none()
-            queryset = queryset.filter(
-                id__in=matches,
+            if 'environment' in tags:
+                # TODO: This should probably just overwrite the existing tag,
+                # rather than asserting on it, but...?
+                assert Environment.objects.get(
+                    projects=project,
+                    name=tags.pop('environment'),
+                ).id == environment.id
+
+            group_matches = set(
+                QuerySetBuilder({
+                    'first_release': CallbackCondition(
+                        lambda queryset, version: queryset.extra(
+                            where=[
+                                '{} = {}'.format(
+                                    get_sql_column(GroupEnvironment, 'first_release_id'),
+                                    get_sql_column(Release, 'id'),
+                                ),
+                                '{} = %s'.format(
+                                    get_sql_column(Release, 'organization'),
+                                ),
+                                '{} = %s'.format(
+                                    get_sql_column(Release, 'version'),
+                                ),
+                            ],
+                            params=[project.organization_id, version],
+                            tables=[Release._meta.db_table],
+                        ),
+                    ),
+                    'times_seen': CallbackCondition(
+                        # This condition represents the exact number of times
+                        # that an issue has been seen in an environment. Since
+                        # an issue can't be seen in an environment more times
+                        # than the issue was seen overall, we can safely
+                        # exclude any groups that don't have at least that many
+                        # events.
+                        lambda queryset, times_seen: queryset.exclude(
+                            times_seen__lt=times_seen,
+                        ),
+                        destructive=False,
+                    ),
+                    'times_seen_lower': CallbackCondition(
+                        # This condition represents the lower threshold for the
+                        # number of times an issue has been seen in an
+                        # environment. Since an issue can't be seen in an
+                        # environment more times than the issue was seen
+                        # overall, we can safely exclude any groups that
+                        # haven't met that threshold.
+                        lambda queryset, times_seen: queryset.exclude(
+                            times_seen__lt=times_seen,
+                        ),
+                        destructive=False,
+                    ),
```

Comment:
Please double check my logic on the filters after this comment to make sure they make sense, it's a little complicated but I think it's a worthwhile optimization for getting more accurate results when using date filters.